File: kakao.app/js/ked/view/base/input.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}}

var input

import kxk from "../../../kxk.js"
let kseg = kxk.kseg

import theme from "../../theme/theme.js"

import editor from "../../edit/editor.js"


input = (function ()
{
    _k_.extend(input, editor)
    function input (screen, name)
    {
        this["current"] = this["current"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onMouseLeave"] = this["onMouseLeave"].bind(this)
        input.__super__.constructor.call(this,screen,name,['brckts','replex','unype'])
        this.setColor('selection_line',this.color.selection)
        this.setColor('bg',theme.quicky.bg)
        this.setColor('bg_blur',theme.quicky.bg)
        this.setColor('bg_focus',theme.editor.bg)
    }

    input.prototype["hasFocus"] = function ()
    {
        return this.state.hasFocus
    }

    input.prototype["onMouseLeave"] = function (event)
    {
        this.emit('action','cancel')
        return input.__super__.onMouseLeave.call(this,event)
    }

    input.prototype["onMouse"] = function (event)
    {
        if (this.hidden())
        {
            return
        }
        return input.__super__.onMouse.call(this,event)
    }

    input.prototype["current"] = function ()
    {
        return kseg.str(this.state.s.lines[0])
    }

    input.prototype["set"] = function (text)
    {
        return this.state.loadLines([text])
    }

    input.prototype["selectAll"] = function ()
    {
        this.state.selectLine(0)
        return this.state.moveCursors('eol')
    }

    input.prototype["onKey"] = function (key, event)
    {
        var before, sr

        if (!this.hasFocus())
        {
            return
        }
        switch (event.combo)
        {
            case 'return':
                this.emit('action','submit',this.current())
                return true

            case 'up':
            case 'down':
                this.emit('action',event.combo)
                return false

            case 'right':
                if (this.state.mainCursor()[0] >= this.current().length)
                {
                    this.emit('action','right')
                    return true
                }
                break
            case 'esc':
                this.emit('action','cancel')
                return true

        }

        before = this.current()
        sr = input.__super__.onKey.call(this,key,event)
        if (before !== this.current())
        {
            this.emit('action','change',this.current())
            return true
        }
        return sr
    }

    input.prototype["draw"] = function ()
    {
        if (this.hidden() || this.collapsed())
        {
            return
        }
        this.setColor('bg',(this.hasFocus() ? this.color.bg_focus : this.color.bg_blur))
        this.setColor('empty',this.color.bg)
        return input.__super__.draw.call(this)
    }

    return input
})()

export default input;

File: kakao.app/love/lua/view/base/input.lua
--[[
    ███  ███   ███  ████████   ███   ███  █████████ 
    ███  ████  ███  ███   ███  ███   ███     ███    
    ███  ███ █ ███  ████████   ███   ███     ███    
    ███  ███  ████  ███        ███   ███     ███    
    ███  ███   ███  ███         ███████      ███    

    simple text input field
    used in inputchoice to filter choices
--]]

editor = require "edit.editor"


local input = class("input", editor)
    


function input:init(name) 
        editor.init(self, name, array('brckts', 'replex', 'unype'))
        
        self:setColor('selection_line', self.color.selection)
        self:setColor('bg', theme.quicky.bg)
        self:setColor('bg_blur', theme.quicky.bg)
        self:setColor('bg_focus', theme.editor.bg)
        return self
    end

-- ████████   ███████    ███████  ███   ███   ███████
-- ███       ███   ███  ███       ███   ███  ███     
-- ██████    ███   ███  ███       ███   ███  ███████ 
-- ███       ███   ███  ███       ███   ███       ███
-- ███        ███████    ███████   ███████   ███████ 


function input:hasFocus() 
    return self.state.hasFocus
    end


function input:onMouseLeave(event) 
        self:emit('action', 'cancel')
        return editor.onMouseLeave(self, event)
    end


function input:onMouse(event) 
        if self:hidden() then return end
        
        return editor.onMouse(self, event)
    end


function input:current() 
        if (valid(self.state.s.lines) and (#self.state.s.lines > 0)) then 
            print("CURRENT " .. tostring(#self.state.s.lines) .. "")
            return self.state.s.lines[1]:arr():str()
        end
        
        return ""
    end


function input:set(text) 
        return self.state:loadLines(array(text))
    end


function input:selectAll() 
        self.state:selectLine(0)
        return self.state:moveCursors('eol')
    end

-- ███   ███  ████████  ███   ███
-- ███  ███   ███        ███ ███ 
-- ███████    ███████     █████  
-- ███  ███   ███          ███   
-- ███   ███  ████████     ███   


function input:onKey(key, event) 
        if not self:hasFocus() then return end
        
        if (event.combo == 'return') then 
                self:emit('action', 'submit', self:current())
                return true
        elseif (event.combo == 'up') or (event.combo == 'down') then 
                self:emit('action', event.combo)
                return false
        elseif (event.combo == 'right') then 
                if (self.state.mainCursor()[0] >= #self:current()) then 
                    self:emit('action', 'right')
                    return true
                end
        elseif (event.combo == 'esc') then 
                self:emit('action', 'cancel')
                return true
        end
        
        local before = self:current()
        
        local sr = editor.onKey(self, key, event)
        
        if (before ~= self:current()) then 
            self:emit('action', 'change', self:current())
            return true
        end
        
        return sr
    end

-- ███████    ████████    ███████   ███   ███
-- ███   ███  ███   ███  ███   ███  ███ █ ███
-- ███   ███  ███████    █████████  █████████
-- ███   ███  ███   ███  ███   ███  ███   ███
-- ███████    ███   ███  ███   ███  ██     ██


function input:draw() 
        if (self:hidden() or self:collapsed()) then return end
        
        local bg = (function () 
    if self:hasFocus() then 
    return self.color.bg_focus else 
    return self.color.bg_blur
             end
end)()
        self:setColor('bg', bg)
        self:setColor('empty', self.color.bg)
        
        return editor.draw(self)
    end

return input